modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PartitionMeta.java

```diff
@@ -23,258 +23,44 @@
 import java.lang.invoke.MethodHandles;
 import java.lang.invoke.VarHandle;
 import java.util.UUID;
+import org.apache.ignite.internal.pagememory.persistence.PartitionMeta.PartitionMetaSnapshot;
 import org.apache.ignite.internal.pagememory.persistence.io.PartitionMetaIo;
-import org.apache.ignite.internal.tostring.S;
 import org.jetbrains.annotations.Nullable;
-import org.jetbrains.annotations.TestOnly;
 
 /**
  * Partition meta information.
  */
-public class PartitionMeta {
-    private static final VarHandle PAGE_COUNT;
+public abstract class PartitionMeta<S extends PartitionMetaSnapshot<?>> {
```

Comment:
It seems that parametrization costs us too much here. If I'm not mistaken, there are just a couple of places where we benefit from the introduced generics:

1. `PersistentPageMemoryMvPartitionStorage#flush()` where we need to know the type of the snapshot
2. `CheckpointPagesWriter#writePartitionMeta()` where we need to know the type of the PageIo to write to

In the first case, we know the actual type statically at the call site (as `meta` has a concrete type). `PartitionMeta#metaSnapshot()` could be overwidden in `StoragePartitionMeta` specifying `StoragePartitionMetaSnapshot` as the return type, this would be enough, no generic is needed.

In the second case, we actually lose type information due to raw types, and we have an implicit type cast. We could just make `PartitionMetaSnapshot#writeTo()` parameter to be of abstract type (`PageIo`) and cast it explicitly in the implementation method. Just one cast is not too bad.

On the other hand, the approach with parametrization requires to parameterize quite a few types all over pagememory code (and to make it work, we need to parameterize even more, to stop having those raw-typed variables).

So my suggestion is to drop the type parameterization approach.

Reply:
done